src/ucp/core/ucp_ep.c

```diff
@@ -1285,17 +1345,29 @@ void ucp_ep_cleanup_lanes(ucp_ep_h ep)
 
     for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
         uct_ep = uct_eps[lane];
-        if (uct_ep == NULL) {
+        if ((uct_ep == NULL) || ucp_is_uct_ep_failed(uct_ep)) {
             continue;
         }
 
+        ++num_closed_lanes;
         ucs_debug("ep %p: pending & destroy uct_ep[%d]=%p", ep, lane, uct_ep);
         uct_ep_pending_purge(uct_ep, ucp_destroyed_ep_pending_purge, ep);
         /* coverity wrongly resolves ucp_failed_tl_ep's no-op EP destroy
          * function to 'ucp_proxy_ep_destroy' */
         /* coverity[incorrect_free] */
         uct_ep_destroy(uct_ep);
     }
+
+    if (num_closed_lanes > 0) {
+        /* Do purging of the requests only if UCT endpoints were closed here,
+         * otherwise - it means that purging of the requests should be done
+         * when the discarding of all lanes will be completed */
+        ucp_ep_reqs_purge(ep, UCS_ERR_CANCELED);
```

Comment:
what if some uct_eps were set to failed and some were not? All requests will be purged here then

Reply:
it shouldn't happen actually.
they should be all set to failed or vice versa

Reply:
maybe add some assertion to guarantee it?

Reply:
would be good

Reply:
done

Reply:
why need to purge here? isn't it already done as part of discard lanes?